### PR TITLE
bugfix: SQLIntegrityConstraintViolationException capture incorrectly when inserting a globallock

### LINKED ISSUE #6822: 已经关闭的问题又出现了:Duplicate entry for key 'lock_table.PRIMARY'
[#4872](https://github.com/apache/incubator-seata/issues/4872)

我使用的是2.0.0的版本, 但出现了一样的问题, 先删后增报错, 尝试关闭lock_table的主键, 会发现两条row_key相同,branch_id不同的数据

user_role表就两个字段: user_id, role_id
删除sql: delete from user_role where user_id = 1
插入sql: insert into user_role (user_id, role_id) values (1, 2)

### PR DESCRIPTION
**Pull request by changeAtLater**
- [ ] I have registered the PR [changes](../changes).

### Ⅱ. Does this pull request fix one issue?

fixes #6822

### CODE REVIEW

File: server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java [hunk 1]
```diff
@@ -16,11 +16,7 @@
  */
 package org.apache.seata.server.storage.db.lock;
 
-import java.sql.Connection;
-import java.sql.PreparedStatement;
-import java.sql.ResultSet;
-import java.sql.SQLException;
-import java.sql.SQLIntegrityConstraintViolationException;
+import java.sql.*;
```
Comment (funky-eyes) [inline server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java]: 不要使用*引入
Do not use * introduce

File: server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java [hunk 2]
```diff
@@ -372,10 +368,27 @@ protected boolean doAcquireLocks(Connection conn, List<LockDO> lockDOs) throws S
...
+                if (cause instanceof SQLIntegrityConstraintViolationException) {
+                    return false;
+                }
+                throw e;
+            }
+            SQLException nextException = e.getNextException();
+            if (nextException == null) {
+                throw e;
+            }
+            while (nextException != null) {
```
Comment (funky-eyes) [inline server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java]: SQLIntegrityConstraintViolationException 有可能会被这么多层包裹吗？
Is it possible for SQLIntegrityConstraintViolationException to be wrapped in so many layers?

**Review by funky-eyes — COMMENTED** [review]
https://github.com/apache/incubator-seata/blob/2.x/changes/zh-cn/2.x.md
https://github.com/apache/incubator-seata/blob/2.x/changes/en-us/2.x.md
请在上述md文件中登记pr信息和作者信息
Please register pr information and author information in the above md file

File: server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java (hunk 2 above)
Comment (changeAtLater) [reply server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java]: BatchUpdateException存在多个异常时nextException会多层嵌套,我希望判断每个异常,如果存在非主键冲突的异常就抛出.当然,我没有处理过BatchUpdateException,遍历所有异常来判断或许过于极端, 还请你提供建议,谢谢

Comment (funky-eyes) [reply server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java]: > Author

BatchUpdateException中的nextException主要与发出的sql有关，这么多的sql可能触发什么异常，那么nextException就可能有哪些异常，所以需要根据这个sql进行判断，目前这个sql只是一个批量的insert，再结合其只有一个主键作为rowkey，防止脏写，所以我认为他只可能出现SQLIntegrityConstraintViolationException这个nextException，不太可能出现其它的异常。

Comment (changeAtLater) [reply server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java]: ok,那我修改一下

**Comment by funky-eyes** [comment]
```
Caused by: java.sql.BatchUpdateException: Duplicate entry 'jdbc:mysql://127.0.0.1:3306/oa^^^sys_user_role^^^10000002_3' for key 'lock_table.PRIMARY'
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_231]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source) ~[na:1.8.0_231]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source) ~[na:1.8.0_231]
        at java.lang.reflect.Constructor.newInstance(Unknown Source) ~[na:1.8.0_231]
        at com.mysql.cj.util.Util.handleNewInstance(Util.java:192) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.util.Util.getInstance(Util.java:167) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.util.Util.getInstance(Util.java:174) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.exceptions.SQLError.createBatchUpdateException(SQLError.java:224) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeBatchedInserts(ClientPreparedStatement.java:755) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeBatchInternal(ClientPreparedStatement.java:426) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.StatementImpl.executeBatch(StatementImpl.java:800) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.alibaba.druid.pool.DruidPooledPreparedStatement.executeBatch(DruidPooledPreparedStatement.java:565) ~[druid-1.2.6.jar:1.2.6]
        at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.doAcquireLocks(LockStoreDataBaseDAO.java:371) ~[classes!/:1.7.0]
        at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.acquireLock(LockStoreDataBaseDAO.java:187) ~[classes!/:1.7.0]
        ... 24 common frames omitted
Caused by: java.sql.SQLIntegrityConstraintViolationException: Duplicate entry 'jdbc:mysql://127.0.0.1:3306/oa^^^sys_user_role^^^10000002_3' for key 'lock_table.PRIMARY'
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:117) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:953) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1098) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1046) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1371) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        at com.mysql.cj.jdbc.ClientPreparedStatement.executeBatchedInserts(ClientPreparedStatement.java:716) ~[mysql-connector-java-8.0.27.jar:8.0.27]
        ... 29 common frames omitted
<==
```
根据这个堆栈，其实可以看到批量插入的时候，nextException是SQLIntegrityConstraintViolationException，如果存在其他异常，可能是数据库之类的异常，这种情况应该属于sqlexception，也不需要额外处理，直接抛出去即可

**Comment by changeAtLater** [comment]
> ```
> Caused by: java.sql.BatchUpdateException: Duplicate entry 'jdbc:mysql://127.0.0.1:3306/oa^^^sys_user_role^^^10000002_3' for key 'lock_table.PRIMARY'
>         at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_231]
>         at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source) ~[na:1.8.0_231]
>         at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source) ~[na:1.8.0_231]
>         at java.lang.reflect.Constructor.newInstance(Unknown Source) ~[na:1.8.0_231]
>         at com.mysql.cj.util.Util.handleNewInstance(Util.java:192) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.util.Util.getInstance(Util.java:167) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.util.Util.getInstance(Util.java:174) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.exceptions.SQLError.createBatchUpdateException(SQLError.java:224) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.ClientPreparedStatement.executeBatchedInserts(ClientPreparedStatement.java:755) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.ClientPreparedStatement.executeBatchInternal(ClientPreparedStatement.java:426) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.StatementImpl.executeBatch(StatementImpl.java:800) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.alibaba.druid.pool.DruidPooledPreparedStatement.executeBatch(DruidPooledPreparedStatement.java:565) ~[druid-1.2.6.jar:1.2.6]
>         at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.doAcquireLocks(LockStoreDataBaseDAO.java:371) ~[classes!/:1.7.0]
>         at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.acquireLock(LockStoreDataBaseDAO.java:187) ~[classes!/:1.7.0]
>         ... 24 common frames omitted
> Caused by: java.sql.SQLIntegrityConstraintViolationException: Duplicate entry 'jdbc:mysql://127.0.0.1:3306/oa^^^sys_user_role^^^10000002_3' for key 'lock_table.PRIMARY'
>         at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:117) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:953) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1098) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1046) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1371) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         at com.mysql.cj.jdbc.ClientPreparedStatement.executeBatchedInserts(ClientPreparedStatement.java:716) ~[mysql-connector-java-8.0.27.jar:8.0.27]
>         ... 29 common frames omitted
> <==
> ```
> 
> 根据这个堆栈，其实可以看到批量插入的时候，nextException是SQLIntegrityConstraintViolationException，如果存在其他异常，可能是数据库之类的异常，这种情况应该属于sqlexception，也不需要额外处理，直接抛出去即可

我用mysql测试时,SQLIntegrityConstraintViolationException在cause中, 但我写单元测试时发现, h2的SQLIntegrityConstraintViolationException在nextException中

**Comment by changeAtLater** [comment]
nextException并不会在日志中打印出来,因此我最后选择抛出nextException,可以更好的知道是什么异常

File: server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java [hunk 3]
```diff
@@ -372,10 +373,23 @@ protected boolean doAcquireLocks(Connection conn, List<LockDO> lockDOs) throws S
...
+                }
+                throw e;
+            }
+            SQLException nextException = e.getNextException();
+            if (nextException == null) {
+                throw e;
+            } else if (nextException instanceof SQLIntegrityConstraintViolationException) {
+                return false;
+            }
+            throw nextException;
```
Comment (funky-eyes) [inline server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java:392]: 我认为这一块代码都是不需要的
I don't think this piece of code is necessary

Comment (changeAtLater) [reply server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java:392]: 那单元测试也可以删了,测试不能通过,另外我只试过mysql和h2,不能保证其他数据库的异常都在cause中,如果你坚持的话,那我就把这段和单元测试都删了

Comment (changeAtLater) [reply server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java:392]: mysql:
image
h2:
image

Comment (funky-eyes) [reply server/src/main/java/org/apache/seata/server/storage/db/lock/LockStoreDataBaseDAO.java:392]: 根据h2的结果来看,似乎这段代码是有意义的,那就按这样留着吧

**Comment by funky-eyes** [comment]
Error:  Failures: 
Error:    DataBaseLockStoreDAOTest.tes_unLocks:178 expected: <true> but was: <false>
Error:    DataBaseLockStoreDAOTest.test_acquireLocks:100 expected: <true> but was: <false>
Error:    DataBaseLockStoreDAOTest.test_isLockable_can:226 expected: <true> but was: <false>
Error:    DataBaseLockStoreDAOTest.test_re_acquireLocks:138 expected: <true> but was: <false>
[INFO] 
Error:  Tests run: 184, Failures: 4, Errors: 0, Skipped: 3

**Review by funky-eyes — APPROVED** [review]
LGTM

**Comment by funky-eyes** [comment]
Thank you for your contributions to Seata. If possible, please send your Dingtalk contact information to my email, and I will invite you to join the Seata developer group.
[jianbin@apache.org](mailto:jianbin@apache.org)

**Pull request merged**